MovingBlocks/Terasology
# Upgrading LWJGL to version 3

### Contains

LWJGL change:
 * Windowing system - GLFW
 * Input system - GLFW
 * GL methods, which use buffers (+suffix)
 * Some GL Extensions moved

We can replace jinput library to GLFW for joystics(controllers) after merge.

### Related PRs

https://github.com/MovingBlocks/TeraNUI/pull/24 - after extraction NUI
https://github.com/MovingBlocks/SplashScreen/pull/4 - GLFW splash screen's classes.  (Old Swing screen crashes LWJGL3's OpenGL initialization - deep side effect)
https://github.com/Terasology/CoreRendering/pull/15 - Update CoreRendering for lwjgl3


### How to test

Checkout https://github.com/Terasology/CoreRendering/pull/15.
Run and play as usual, check :)

Platforms:
- [x] - Linux(X)
- [ ] - ~~Linux(Wayland) - non goal, GLFW provides support for Wayland~~   LWJGL3 haven't proper code. possibly in next release

Checklist:
 * Input system:
    * Keyboard:
        * keys working, keybindings working
        * chat and console receive chars as usual  - your language work, special symbols work, uppercase correct
    * Mouse:
        * position in NUI work correctly
        * scroll works
        * mouse acceleration  works correctly
 * Windowing system:
    * Windows resizing works
    * Fullscreen works
    * Pseudo fullscreen works
 * Rendering works, not glitches
 * OpenVR works (I hope)
 * Natives publishes in distribution
 * Check switchable graphics on MacOS


[bonus] measure performance on discrete graphics

### Outstanding before merging

- [x] Return ```SkySphere```, LWJGL 3 lost ```glu``` packages
- [x] Merge GLFW's char and key callbacks products... or separate NUIs key events
      Its works separatelly now. Generates two independent ```KeyboardEvent``` streams. 
      [!] You are moving when type in chat or console in-game!
- [x] Repair type adapter for configuration (Resolution not saves)
- [x] ~~Repair ```LwjglPortlet``` (for TeraEd, or use another way)~~
- [x] Repair framelimit
- [x] Fix ```Resolution``` in settings gui (used Object's toString)
- [x] Expose LWJGL 3 to modules
- [x] Saving window position
- [x] Refactor Debug/Error outputs for LWJGL 3
- [x] Overall refactoring
- [x] Replace SwingSplashScreen with LWJGL 3 solution (GLFW+ LWJGL's opengl).
- [x] Merge https://github.com/MovingBlocks/SplashScreen/pull/4
